pkg/apis/resource/validation/validation.go

```diff
@@ -592,23 +595,61 @@ func validateResourceSliceSpec(spec, oldSpec *resource.ResourceSliceSpec, fldPat
 	if spec.AllNodes {
 		setFields = append(setFields, "`allNodes`")
 	}
+	if spec.PerDeviceNodeSelection {
+		setFields = append(setFields, "`perDeviceNodeSelection`")
+	}
 	switch len(setFields) {
 	case 0:
-		allErrs = append(allErrs, field.Required(fldPath, "exactly one of `nodeName`, `nodeSelector`, or `allNodes` is required"))
+		allErrs = append(allErrs, field.Required(fldPath, "exactly one of `nodeName`, `nodeSelector`, `allNodes`, `perDeviceNodeSelection` is required"))
 	case 1:
 	default:
 		allErrs = append(allErrs, field.Invalid(fldPath, fmt.Sprintf("{%s}", strings.Join(setFields, ", ")),
-			"exactly one of `nodeName`, `nodeSelector`, or `allNodes` is required, but multiple fields are set"))
+			"exactly one of `nodeName`, `nodeSelector`, `allNodes`, `perDeviceNodeSelection` is required, but multiple fields are set"))
 	}
 
-	allErrs = append(allErrs, validateSet(spec.Devices, resource.ResourceSliceMaxDevices, validateDevice,
+	capacityPoolNames := gatherCapacityPoolNames(spec.CapacityPools)
+	allErrs = append(allErrs, validateSet(spec.Devices, resource.ResourceSliceMaxDevices,
+		func(device resource.Device, fldPath *field.Path) field.ErrorList {
+			return validateDevice(device, fldPath, capacityPoolNames, spec.PerDeviceNodeSelection)
+		},
 		func(device resource.Device) (string, string) {
 			return device.Name, "name"
 		}, fldPath.Child("devices"))...)
 
+	allErrs = append(allErrs, validateSet(spec.CapacityPools, resource.ResourceSliceMaxCapacityPools,
+		func(capacityPool resource.CapacityPool, fldPath *field.Path) field.ErrorList {
+			return validateCapacityPool(capacityPool, fldPath)
+		},
```

Comment:
Nit:

```suggestion
		validateCapacityPool,
```

Brought to you by your friendly linter:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/130764/pull-kubernetes-linter-hints/1901323231734796288
```
pkg/apis/resource/validation/validation.go:620:3: unlambda: replace `func(capacityPool resource.CapacityPool, fldPath *field.Path) field.ErrorList {
	return validateCapacityPool(capacityPool, fldPath)
}` with `validateCapacityPool` (gocritic)
		func(capacityPool resource.CapacityPool, fldPath *field.Path) field.ErrorList {
		^
}
```